src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/service/PartBRecordService.kt

```diff
@@ -0,0 +1,112 @@
+package uk.gov.justice.digital.hmpps.managerecallsapi.service
+
+import dev.forkhandles.result4k.Failure
+import dev.forkhandles.result4k.Result
+import dev.forkhandles.result4k.Success
+import dev.forkhandles.result4k.map
+import dev.forkhandles.result4k.mapFailure
+import dev.forkhandles.result4k.valueOrNull
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.stereotype.Service
+import uk.gov.justice.digital.hmpps.managerecallsapi.controller.PartBRecordController.PartBRequest
+import uk.gov.justice.digital.hmpps.managerecallsapi.db.DocumentCategory
+import uk.gov.justice.digital.hmpps.managerecallsapi.db.DocumentCategory.OASYS_RISK_ASSESSMENT
+import uk.gov.justice.digital.hmpps.managerecallsapi.db.DocumentCategory.PART_B_EMAIL_FROM_PROBATION
+import uk.gov.justice.digital.hmpps.managerecallsapi.db.DocumentCategory.PART_B_RISK_REPORT
+import uk.gov.justice.digital.hmpps.managerecallsapi.db.PartBRecord
+import uk.gov.justice.digital.hmpps.managerecallsapi.db.PartBRecordRepository
+import uk.gov.justice.digital.hmpps.managerecallsapi.db.RecallRepository
+import uk.gov.justice.digital.hmpps.managerecallsapi.documents.toBase64DecodedByteArray
+import uk.gov.justice.digital.hmpps.managerecallsapi.domain.DocumentId
+import uk.gov.justice.digital.hmpps.managerecallsapi.domain.FileName
+import uk.gov.justice.digital.hmpps.managerecallsapi.domain.PartBRecordId
+import uk.gov.justice.digital.hmpps.managerecallsapi.domain.RecallId
+import uk.gov.justice.digital.hmpps.managerecallsapi.domain.UserId
+import uk.gov.justice.digital.hmpps.managerecallsapi.domain.random
+import java.time.OffsetDateTime
+import javax.transaction.Transactional
+
+@Service
+class PartBRecordService(
+  @Autowired private val partBRecordRepository: PartBRecordRepository,
+  @Autowired private val recallRepository: RecallRepository,
+  @Autowired private val documentService: DocumentService,
+) {
+
+  @Transactional
+  fun createRecord(
+    recallId: RecallId,
+    currentUserId: UserId,
+    request: PartBRequest
+  ): Result<PartBRecordId, List<Pair<DocumentCategory, FileName>>> =
+    recallRepository.getByRecallId(recallId).let { recall ->
+      // TODO: Move validation to DTO?  Even then, cross-field validation requires custom code.
+      //  As these properties are mutually dependent an alternative would be to combine them as a single object in the DTO
+      val oasysName = request.oasysFileName
+      val oasysContent = request.oasysFileContent
+      if ((oasysName == null && !oasysContent.isNullOrBlank()) || (oasysContent.isNullOrBlank() && oasysName != null)) {
+        throw IllegalArgumentException("Both fileName and fileContent of OASYS must be supplied as non-blank or neither")
+      }
```

Comment:
should/could this be on the record itself?

Reply:
I have a // TODO above it to consider how else to apply validation but maybe swings and roundabouts.

Reply:
Only the document Id is on the record so there's no scope there for validating name + content together.  I've added this to the // TODO: [PUD-1626](https://dsdmoj.atlassian.net/browse/PUD-1626) Cross-property Request validation: move to DTOs or restructure Request classes?